### DISCUSSION TITLE
Compare Nickel and KCL?

### DISCUSSION THREAD

**Discussion by johnnyutahh** [tweag/nickel#1582] [Q&A] [answered]
The KCL docs compare KCL with Nickel:

https://kcl-lang.io/docs/user_docs/getting-started/intro/#vs-nickel

Can anyone here (at Nickel) take a pass at comparing Nickel and KCL?

**Comment by jneem** [comment, marked as answer]
I looked into KCL a little today. There are several similarities: both languages have user-extensible runtime checks (contracts in nickel; schemas/protocols in KCL), mergeable records, and static types.

The KCL typesystem feels more nominal and object-oriented than nickel's:
- in KCL you specify the name of the schema when you're writing out the object that's supposed to conform to it; in nickel, you can write out a record first and then apply the contract at some later point
- in KCL, schema inheritance and mixins are written explicitly; in nickel you'd apply multiple contracts, and maybe use row polymorphism instead of sub-schemas

But I think the big difference is in the approach to laziness. AFAICT, none of KCL's runtime checks are lazy and so you can't do the thing (which is intended to be idiomatic in nickel) where you apply a contract that isn't fully satisfied "yet" because it will be satisfied after merging in some other records "later"

**Reply by johnnyutahh** [reply]
This is tremendously helpful, thank you @jneem. I really appreciate your taking the time to evaluate!
